# CAPSTONE PROJECT

### Importing Necessary Libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### Obtaining the Data¶

Firstly, we need a list of neighbourhoods of Toronto. For the Toronto neighborhood data, a Wikipedia page exists :https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M
BeautifulSoup and requests packages will be used for website scrapping. 

In [2]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text
from bs4 import BeautifulSoup
soup = BeautifulSoup(website_url,'html.parser')
neighborhoods_data = soup.find('table',{'class':'wikitable sortable'})
neighborhoods_data_str = str(neighborhoods_data)
df = pd.read_html(neighborhoods_data_str)[0]
df = df.rename(columns = {'Postcode':'PostalCode','Neighbourhood':'Neighborhood'})
df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

### Data Cleaning

There should be done 3 corrections in the dataset. 
1. In Brough column, the rows which are Not assigned will be removed.
2. Postal code areas which have more than one neighbourhood will be combined into one row.
3. In Neighbourhood column, the rows which are Not assigned will be changed to the brough name.

In [3]:
# Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
df.drop(df[df['Borough']=="Not assigned"].index,axis=0, inplace=True)
#More than one neighborhood can exist in one postal code area. 
df1 = df.groupby("PostalCode", as_index=False).agg(lambda x:','.join(set(x)))
# If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. 
df1.loc[df1['Neighborhood']=="Not assigned",'Neighborhood']=df1.loc[df1['Neighborhood']=="Not assigned",'Borough']
df1.head()

PostalCode      Borough                          Neighborhood
0        M1B  Scarborough                         Malvern,Rouge
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union
2        M1E  Scarborough       Guildwood,Morningside,West Hill
3        M1G  Scarborough                                Woburn
4        M1H  Scarborough                             Cedarbrae

In [4]:
df1.shape

(103, 3)

The dataset has 103 rows and 3 columns.

In order to get venue data and plot a map, the coordinates of each neighnourhood is required. It is obtained from a csv file :http://cocl.us/Geospatial_data

In [5]:
# Read a file that includes coordinates of postal codes 
coordinates = pd.read_csv("/home/lpy/Downloads/Geospatial_Coordinates.csv")
coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [6]:
# Adding 'Latitude' and 'Longitude'columns to our dataframe
df1['Latitude']=coordinates['Latitude'].values
df1['Longitude']=coordinates['Longitude'].values
df1.head()

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Malvern,Rouge  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

Since the project focuses on the Toronto only, the data will be filtered.

In [7]:
df2 = df1[df1['Borough'].str.contains("Toronto")].reset_index(drop=True)
df2.head()

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    Riverdale,The Danforth West  43.679557   
2        M4L     East Toronto  India Bazaar,The Beaches West  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

### Exploratory Data Analysis

#### Neighbourhoods of Toronto Map

geopy library is used to get the latitude and longitude values of Toronto.

In [8]:
address = 'Toronto'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [9]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(df2['Latitude'], df2['Longitude'], df2['Borough'], df2['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

#### Explore Neighborhoods in Toronto

Foursquare API will be used to explore venues in the neighbourhoods in Toronto. Top 100 venues which are within a radius of 1000 meters will be requested from Foursquare API.

In [10]:
# Define Foursquare Credentials and Version
CLIENT_ID = 'ULJAWACSBIL41ZFLQZJR4EO4P3KDHXFBIMGWHPVBSZIRC5L1' # your Foursquare ID
CLIENT_SECRET = 'SSZVJPBVK4FVV55M4WL42QWVOTPPKPGCHYIX0F0HCQVKS01L' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

In [11]:
#Create a function that gets the information for all neighbourhoods in Toronto
LIMIT = 100
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [12]:
# run the created function to get venues
toronto_venues = getNearbyVenues(names=df2['Neighborhood'],
                                   latitudes=df2['Latitude'],
                                   longitudes=df2['Longitude']
                                  )


The Beaches
Riverdale,The Danforth West
India Bazaar,The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Summerhill East,Moore Park
Summerhill West,South Hill,Rathnelly,Forest Hill SE,Deer Park
Rosedale
St. James Town,Cabbagetown
Church and Wellesley
Regent Park,Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
King,Richmond,Adelaide
Union Station,Harbourfront East,Toronto Islands
Design Exchange,Toronto Dominion Centre
Victoria Hotel,Commerce Court
Roselawn
Forest Hill North,Forest Hill West
Yorkville,The Annex,North Midtown
University of Toronto,Harbord
Chinatown,Kensington Market,Grange Park
Island airport,King and Spadina,South Niagara,Bathurst Quay,Railway Lands,Harbourfront West,CN Tower
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Trinity,Little Portugal
Exhibition Place,Brockton,Parkdale Village
The Junction South,High Park
Roncesvalles,Parkda

The size of resulting dataframe is as follows:

In [13]:
print(toronto_venues.shape)
toronto_venues.head()

(3084, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

               Venue  Venue Latitude  Venue Longitude  \
0    Tori's Bakeshop       43.672114       -79.290331   
1    The Fox Theatre       43.672801       -79.287272   
2     The Beech Tree       43.680493       -79.288846   
3    Ed's Real Scoop       43.672630       -79.287993   
4  Beaches Bake Shop       43.680363       -79.289692   

                  Venue Category  
0  Vegetarian / Vegan Restaurant  
1            Indie Movie Theater  
2                      Gastropub  
3                 Ice Cream Shop  
4                         Bakery

#### Analyze Each Neighbourhood

In [14]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols = toronto_onehot.columns.tolist()
cols.insert(0, cols.pop(cols.index('Neighborhood')))
toronto_onehot = toronto_onehot.reindex(columns = cols)


toronto_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0  The Beaches                  0               0                  0        0   
1  The Beaches                  0               0                  0        0   
2  The Beaches                  0               0                  0        0   
3  The Beaches                  0               0                  0        0   
4  The Beaches                  0               0                  0        0   

   Airport Lounge  American Restaurant  Amphitheater  Animal Shelter  \
0               0                    0             0               0   
1               0                    0             0               0   
2               0                    0             0               0   
3               0                    0             0               0   
4               0                    0             0               0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Dealership  Auto Garage  \
0                 0                   0                0            0   
1                 0                   0                0            0   
2                 0                   0                0            0   
3                 0                   0                0            0   
4                 0                   0                0            0   

   BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0          0           0       0     0    0               0                 0   
1          0           0       0     0    0               0                 0   
2          0           0       0     0    0               0                 0   
3          0           0       0     0    0               0                 0   
4          0           0       1     0    0               0                 0   

   Basketball Stadium  Beach  Beach Bar  Beer Bar  Beer Store  \
0                   0      0          0         0           0   
1                   0      0          0         0           0   
2                   0      0          0         0           0   
3                   0      0          0         0           0   
4                   0      0          0         0           0   

   Belgian Restaurant  Bike Shop  Bistro  Board Shop  Bookstore  Boutique  \
0                   0          0       0           0          0         0   
1                   0          0       0           0          0         0   
2                   0          0       0           0          0         0   
3                   0          0       0           0          0         0   
4                   0          0       0           0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Bus Station  \
0                0       0             0              0            0   
1                0       0             0              0            0   
2                0       0             0              0            0   
3                0       0             0              0            0   
4                0       0             0              0            0   

   Bus Stop  Business Servi

In [15]:
toronto_onehot.shape

(3084, 275)

Next, rows will be grouped by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [16]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                        Berczy Park                0.0   
1  Business Reply Mail Processing Centre 969 Eastern                0.0   
2                                 Central Bay Street                0.0   
3            Chinatown,Kensington Market,Grange Park                0.0   
4                                           Christie                0.0   

   Adult Boutique  Afghan Restaurant  Airport  Airport Lounge  \
0             0.0                0.0      0.0             0.0   
1             0.0                0.0      0.0             0.0   
2             0.0                0.0      0.0             0.0   
3             0.0                0.0      0.0             0.0   
4             0.0                0.0      0.0             0.0   

   American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
0             0.010000           0.0             0.0           0.0       0.0   
1             0.020833           0.0             0.0           0.0       0.0   
2             0.010000           0.0             0.0           0.0       0.0   
3             0.000000           0.0             0.0           0.0       0.0   
4             0.010000           0.0             0.0           0.0       0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0.02        0.00             0.000000               0.0   
1         0.00        0.00             0.020833               0.0   
2         0.01        0.01             0.030000               0.0   
3         0.02        0.00             0.010000               0.0   
4         0.01        0.00             0.000000               0.0   

   Athletics & Sports  Auto Dealership  Auto Garage  BBQ Joint  Bagel Shop  \
0                 0.0              0.0          0.0   0.010000        0.01   
1                 0.0              0.0          0.0   0.020833        0.00   
2                 0.0              0.0          0.0   0.000000        0.00   
3                 0.0              0.0          0.0   0.000000        0.01   
4                 0.0              0.0          0.0   0.000000        0.00   

     Bakery  Bank       Bar  Baseball Field  Baseball Stadium  \
0  0.030000  0.00  0.000000             0.0               0.0   
1  0.020833  0.00  0.020833             0.0               0.0   
2  0.010000  0.00  0.020000             0.0               0.0   
3  0.040000  0.00  0.070000             0.0               0.0   
4  0.010000  0.01  0.020000             0.0               0.0   

   Basketball Stadium     Beach  Beach Bar  Beer Bar  Beer Store  \
0                0.01  0.010000        0.0      0.04         0.0   
1                0.00  0.020833        0.0      0.00         0.0   
2                0.00  0.000000        0.0      0.01         0.0   
3                0.00  0.000000        0.0      0.01         0.0   
4                0.00  0.000000        0.0      0.00         0.0   

   Belgian Restaurant  Bike Shop    Bistro  Board Shop  Bookstore  Boutique  \
0                0.01        0.0  0.010000         0.0       0.00       0.0   
1                0.00        0.0  0.020833         0.0       0.00       0.0   
2                0.00        0.0  0.000000         0.0       0.01       0.0   
3                0.01        0.0  0.000000         0.0       0.00       0.0   
4                0.00        0.0  0.000000         0.0       0.00       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                   0.0        0.020000   0.0000         0.00   
1                   0.0        0.020833   0.0625         0.00   
2                   0.0        0.020000   0.0000         0.00   
3                   0.0        0.000000   0.0100         0.00   
4                   0.0        0.000000   0.0000         0.01   

   Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Bus Station  \
0             0.01     0.0      0.000000       0.000000          0.0   
1             0.00     

Since Coffee Shops are the interest of this project, the dataframe will be filtered to coffee shops only.

In [17]:
cof_shop = toronto_grouped[["Neighborhood","Coffee Shop"]]
cof_shop.head()

Neighborhood  Coffee Shop
0                                        Berczy Park     0.070000
1  Business Reply Mail Processing Centre 969 Eastern     0.083333
2                                 Central Bay Street     0.080000
3            Chinatown,Kensington Market,Grange Park     0.030000
4                                           Christie     0.050000

#### Cluster Neighborhoods

k-means clustering method will be used to analyze the data. The number of clusters, which is k, is determined to be 3. 

In [18]:
# set number of clusters
kclusters = 3

cof_shop_clustering = cof_shop.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(cof_shop_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 2, 2, 0, 0, 1, 2, 0], dtype=int32)

After obtaining the labels, they are added to the data.

In [19]:
# create a copy of coffee shop-neighborhood data
cof_shop_df = cof_shop.copy()
# add cluster labels
cof_shop_df["Cluster Label"] = kmeans.labels_

In [20]:
cof_shop_df.head()

Neighborhood  Coffee Shop  \
0                                        Berczy Park     0.070000   
1  Business Reply Mail Processing Centre 969 Eastern     0.083333   
2                                 Central Bay Street     0.080000   
3            Chinatown,Kensington Market,Grange Park     0.030000   
4                                           Christie     0.050000   

   Cluster Label  
0              0  
1              0  
2              0  
3              2  
4              2

In [21]:
# Add latitude/longitude for each neighborhood
coffee_shop = df2.join(cof_shop_df.set_index('Neighborhood'), on='Neighborhood')

coffee_shop.head()

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    Riverdale,The Danforth West  43.679557   
2        M4L     East Toronto  India Bazaar,The Beaches West  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Coffee Shop  Cluster Label  
0 -79.293031     0.060976              2  
1 -79.352188     0.090000              0  
2 -79.315572     0.053333              2  
3 -79.340923     0.070000              0  
4 -79.388790     0.111111              0

#### Visualization of Resulting Clusters

In [22]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(coffee_shop['Latitude'], coffee_shop['Longitude'], coffee_shop['Neighborhood'], coffee_shop['Cluster Label']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Examine Clusters

Cluster 0

In [23]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 0]

PostalCode           Borough  \
1         M4K      East Toronto   
3         M4M      East Toronto   
4         M4N   Central Toronto   
6         M4R   Central Toronto   
7         M4S   Central Toronto   
8         M4T   Central Toronto   
9         M4V   Central Toronto   
12        M4Y  Downtown Toronto   
16        M5E  Downtown Toronto   
17        M5G  Downtown Toronto   
19        M5J  Downtown Toronto   
21        M5L  Downtown Toronto   
22        M5N   Central Toronto   
23        M5P   Central Toronto   
28        M5W  Downtown Toronto   
29        M5X  Downtown Toronto   
31        M6H      West Toronto   
36        M6S      West Toronto   
37        M7Y      East Toronto   

                                         Neighborhood   Latitude  Longitude  \
1                         Riverdale,The Danforth West  43.679557 -79.352188   
3                                     Studio District  43.659526 -79.340923   
4                                       Lawrence Park  43.728020 -79.388790   
6                                  North Toronto West  43.715383 -79.405678   
7                                          Davisville  43.704324 -79.388790   
8                          Summerhill East,Moore Park  43.689574 -79.383160   
9   Summerhill West,South Hill,Rathnelly,Forest Hi...  43.686412 -79.400049   
12                               Church and Wellesley  43.665860 -79.383160   
16                                        Berczy Park  43.644771 -79.373306   
17                                 Central Bay Street  43.657952 -79.387383   
19    Union Station,Harbourfront East,Toronto Islands  43.640816 -79.381752   
21                      Victoria Hotel,Commerce Court  43.648198 -79.379817   
22                                           Roselawn  43.711695 -79.416936   
23                 Forest Hill North,Forest Hill West  43.696948 -79.411307   
28                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846   
29              First Canadian Place,Underground city  43.648429 -79.382280   
31                        Dovercourt Village,Dufferin  43.669005 -79.442259   
36                                  Swansea,Runnymede  43.651571 -79.484450   
37  Business Reply Mail Processing Centre 969 Eastern  43.662744 -79.321558   

    Coffee Shop  Cluster Label  
1      0.090000              0  
3      0.070000              0  
4      0.111111              0  
6      0.068182              0  
7      0.090000              0  
8      0.084746              0  
9      0.103896              0  
12     0.080000              0  
16     0.070000              0  
17     0.080000              0  
19     0.070000              0  
21     0.080000              0  
22     0.074074              0  
23     0.066667              0  
28     0.090000              0  
29     0.070000              0  
31     0.073529              0  
36     0.093333              0  
37     0.083333              0

In [24]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 0]['Coffee Shop'].mean()

0.0815195586880532

Cluster 1

In [25]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 1]

PostalCode           Borough  \
5         M4P   Central Toronto   
10        M4W  Downtown Toronto   
13        M5A  Downtown Toronto   
27        M5V  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  \
5                                    Davisville North  43.712751 -79.390197   
10                                           Rosedale  43.679563 -79.377529   
13                           Regent Park,Harbourfront  43.654260 -79.360636   
27  Island airport,King and Spadina,South Niagara,...  43.628947 -79.394420   

    Coffee Shop  Cluster Label  
5      0.130000              1  
10     0.130435              1  
13     0.160000              1  
27     0.187500              1

In [26]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 1]['Coffee Shop'].mean()

0.15198369565217393

Cluster 2

In [27]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 2]

PostalCode           Borough                                Neighborhood  \
0         M4E      East Toronto                                 The Beaches   
2         M4L      East Toronto               India Bazaar,The Beaches West   
11        M4X  Downtown Toronto                  St. James Town,Cabbagetown   
14        M5B  Downtown Toronto                     Ryerson,Garden District   
15        M5C  Downtown Toronto                              St. James Town   
18        M5H  Downtown Toronto                      King,Richmond,Adelaide   
20        M5K  Downtown Toronto     Design Exchange,Toronto Dominion Centre   
24        M5R   Central Toronto           Yorkville,The Annex,North Midtown   
25        M5S  Downtown Toronto               University of Toronto,Harbord   
26        M5T  Downtown Toronto     Chinatown,Kensington Market,Grange Park   
30        M6G  Downtown Toronto                                    Christie   
32        M6J      West Toronto                     Trinity,Little Portugal   
33        M6K      West Toronto  Exhibition Place,Brockton,Parkdale Village   
34        M6P      West Toronto                The Junction South,High Park   
35        M6R      West Toronto                       Roncesvalles,Parkdale   

     Latitude  Longitude  Coffee Shop  Cluster Label  
0   43.676357 -79.293031     0.060976              2  
2   43.668999 -79.315572     0.053333              2  
11  43.667967 -79.367675     0.052632              2  
14  43.657162 -79.378937     0.060000              2  
15  43.651494 -79.375418     0.060000              2  
18  43.650571 -79.384568     0.050000              2  
20  43.647177 -79.381576     0.060000              2  
24  43.672710 -79.405678     0.040000              2  
25  43.662696 -79.400049     0.030000              2  
26  43.653206 -79.400049     0.030000              2  
30  43.669542 -79.422564     0.050000              2  
32  43.647927 -79.419750     0.040000              2  
33  43.636847 -79.428191     0.060000              2  
34  43.661608 -79.464763     0.061224              2  
35  43.648960 -79.456325     0.060000              2

In [28]:
coffee_shop.loc[coffee_shop['Cluster Label'] == 2]['Coffee Shop'].mean()

0.05121100078884785

According to the cluster averages, Cluster 2 has the minimum frequency of occurrence of Coffee Shop. So that it is reasonable to make further exploration on those neighborhoods with the label cluster 2 since merket competition is low in those areas.

#### Examine Other Venues in Selected Cluster (Cluster 2)

In [34]:
# make a new dataframe called target that includes cluster 1 data
target = coffee_shop.loc[coffee_shop['Cluster Label'] == 2]
target.shape

(15, 7)

All neighborhoods that are labeled as Cluster 2 assigned in a dataframe to analyze different categories of venues.

In [36]:
target_grouped =toronto_grouped.loc[toronto_grouped['Neighborhood'].isin(target['Neighborhood'])]
target_grouped.head()

Neighborhood  Accessories Store  \
3      Chinatown,Kensington Market,Grange Park               0.00   
4                                     Christie               0.00   
8      Design Exchange,Toronto Dominion Centre               0.00   
10  Exhibition Place,Brockton,Parkdale Village               0.01   
13               India Bazaar,The Beaches West               0.00   

    Adult Boutique  Afghan Restaurant  Airport  Airport Lounge  \
3              0.0                0.0      0.0             0.0   
4              0.0                0.0      0.0             0.0   
8              0.0                0.0      0.0             0.0   
10             0.0                0.0      0.0             0.0   
13             0.0                0.0      0.0             0.0   

    American Restaurant  Amphitheater  Animal Shelter  Antique Shop  Aquarium  \
3                  0.00           0.0             0.0           0.0      0.00   
4                  0.01           0.0             0.0           0.0      0.00   
8                  0.03           0.0             0.0           0.0      0.01   
10                 0.01           0.0             0.0           0.0      0.00   
13                 0.00           0.0             0.0           0.0      0.00   

    Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
3      0.020000         0.0                 0.01          0.000000   
4      0.010000         0.0                 0.00          0.000000   
8      0.010000         0.0                 0.00          0.010000   
10     0.010000         0.0                 0.02          0.000000   
13     0.013333         0.0                 0.00          0.013333   

    Athletics & Sports  Auto Dealership  Auto Garage  BBQ Joint  Bagel Shop  \
3                 0.00              0.0          0.0   0.000000        0.01   
4                 0.00              0.0          0.0   0.000000        0.00   
8                 0.00              0.0          0.0   0.000000        0.00   
10                0.01              0.0          0.0   0.000000        0.00   
13                0.00              0.0          0.0   0.013333        0.00   

      Bakery  Bank       Bar  Baseball Field  Baseball Stadium  \
3   0.040000  0.00  0.070000        0.000000               0.0   
4   0.010000  0.01  0.020000        0.000000               0.0   
8   0.020000  0.00  0.010000        0.000000               0.0   
10  0.030000  0.00  0.040000        0.000000               0.0   
13  0.026667  0.00  0.013333        0.013333               0.0   

    Basketball Stadium  Beach  Beach Bar  Beer Bar  Beer Store  \
3                 0.00   0.00       0.00      0.01         0.0   
4                 0.00   0.00       0.00      0.00         0.0   
8                 0.01   0.00       0.00      0.02         0.0   
10                0.00   0.00       0.01      0.02         0.0   
13                0.00   0.04       0.00      0.00         0.0   

    Belgian Restaurant  Bike Shop  Bistro  Board Shop  Bookstore  Boutique  \
3                 0.01        0.0    0.00    0.000000        0.0       0.0   
4                 0.00        0.0    0.00    0.000000        0.0       0.0   
8                 0.00        0.0    0.01    0.000000        0.0       0.0   
10                0.00        0.0    0.00    0.000000        0.0       0.0   
13                0.00        0.0    0.00    0.013333        0.0       0.0   

    Brazilian Restaurant  Breakfast Spot   Brewery  Bridal Shop  \
3                   0.00        0.000000  0.010000         0.00   
4                   0.00        0.000000  0.000000         0.01   
8                   0.01        0.010000  0.000000         0.00   
10                  0.00        0.010000  0.010000         0.00   
13                  0.00        0.013333  0.026667         0.00   

    Bubble Tea Shop  Buffet  Burger Joint  Burrito Place  Bus Station  \
3              0.00     0.0      0.020000       0.010000          0.0   
4              0.01     0.0

In [37]:
#confirm the new size
target_grouped.shape

(15, 275)

In [38]:
#write a function to sort the venues in descending order.
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [50]:
#create the new dataframe and display the top 4 venues for each neighborhood.
num_top_venues = 4

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
target_venues_sorted = pd.DataFrame(columns=columns)
target_venues_sorted['Neighborhood'] = target_grouped['Neighborhood']

for ind in np.arange(target_grouped.shape[0]):
    target_venues_sorted.iloc[ind, 1:] = return_most_common_venues(target_grouped.iloc[ind, :], num_top_venues)

target_venues_sorted.reset_index(drop=True)

Neighborhood 1st Most Common Venue  \
0      Chinatown,Kensington Market,Grange Park                  Café   
1                                     Christie     Korean Restaurant   
2      Design Exchange,Toronto Dominion Centre                  Café   
3   Exhibition Place,Brockton,Parkdale Village                  Café   
4                India Bazaar,The Beaches West     Indian Restaurant   
5                       King,Richmond,Adelaide                  Café   
6                        Roncesvalles,Parkdale           Coffee Shop   
7                      Ryerson,Garden District           Coffee Shop   
8                               St. James Town                  Café   
9                   St. James Town,Cabbagetown                  Park   
10                                 The Beaches                   Pub   
11                The Junction South,High Park                   Bar   
12                     Trinity,Little Portugal                   Bar   
13               University of Toronto,Harbord                  Café   
14           Yorkville,The Annex,North Midtown                  Café   

            2nd Most Common Venue          3rd Most Common Venue  \
0   Vegetarian / Vegan Restaurant                            Bar   
1                            Café                  Grocery Store   
2                           Hotel                    Coffee Shop   
3                     Coffee Shop                            Bar   
4                     Coffee Shop                           Café   
5                     Coffee Shop                          Hotel   
6                     Pizza Place                            Bar   
7                            Café                 Cosmetics Shop   
8                     Coffee Shop                     Restaurant   
9                            Café            Japanese Restaurant   
10                    Pizza Place                    Coffee Shop   
11                           Café                    Coffee Shop   
12                           Café                         Bakery   
13                         Bakery  Vegetarian / Vegan Restaurant   
14  Vegetarian / Vegan Restaurant             Italian Restaurant   

        4th Most Common Venue  
0          Mexican Restaurant  
1                 Coffee Shop  
2                  Steakhouse  
3      Furniture / Home Store  
4                       Beach  
5                     Theater  
6                        Café  
7   Middle Eastern Restaurant  
8                       Hotel  
9                       Diner  
10                      Beach  
11          Convenience Store  
12                Coffee Shop  
13                        Bar  
14                Coffee Shop

k-means clustering method is used again to analyze the most important venues in the selected areas.

In [51]:
# set number of clusters
kclusters = 3

target_grouped_clustering = target_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(target_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 0, 1, 2, 0, 2, 0, 0, 0], dtype=int32)

In [52]:
# add clustering labels
target_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
target_venues_sorted

Cluster Labels                                Neighborhood  \
3                1     Chinatown,Kensington Market,Grange Park   
4                1                                    Christie   
8                0     Design Exchange,Toronto Dominion Centre   
10               1  Exhibition Place,Brockton,Parkdale Village   
13               2               India Bazaar,The Beaches West   
15               0                      King,Richmond,Adelaide   
20               2                       Roncesvalles,Parkdale   
23               0                     Ryerson,Garden District   
24               0                              St. James Town   
25               0                  St. James Town,Cabbagetown   
31               2                                 The Beaches   
32               1                The Junction South,High Park   
33               1                     Trinity,Little Portugal   
35               1               University of Toronto,Harbord   
37               1           Yorkville,The Annex,North Midtown   

   1st Most Common Venue          2nd Most Common Venue  \
3                   Café  Vegetarian / Vegan Restaurant   
4      Korean Restaurant                           Café   
8                   Café                          Hotel   
10                  Café                    Coffee Shop   
13     Indian Restaurant                    Coffee Shop   
15                  Café                    Coffee Shop   
20           Coffee Shop                    Pizza Place   
23           Coffee Shop                           Café   
24                  Café                    Coffee Shop   
25                  Park                           Café   
31                   Pub                    Pizza Place   
32                   Bar                           Café   
33                   Bar                           Café   
35                  Café                         Bakery   
37                  Café  Vegetarian / Vegan Restaurant   

            3rd Most Common Venue      4th Most Common Venue  
3                             Bar         Mexican Restaurant  
4                   Grocery Store                Coffee Shop  
8                     Coffee Shop                 Steakhouse  
10                            Bar     Furniture / Home Store  
13                           Café                      Beach  
15                          Hotel                    Theater  
20                            Bar                       Café  
23                 Cosmetics Shop  Middle Eastern Restaurant  
24                     Restaurant                      Hotel  
25            Japanese Restaurant                      Diner  
31                    Coffee Shop                      Beach  
32                    Coffee Shop          Convenience Store  
33                         Bakery                Coffee Shop  
35  Vegetarian / Vegan Restaurant                        Bar  
37             Italian Restaurant                Coffee Shop

Cluster 0

In [54]:
target_venues_sorted.loc[target_venues_sorted['Cluster Labels'] == 0]

Cluster Labels                             Neighborhood  \
8                0  Design Exchange,Toronto Dominion Centre   
15               0                   King,Richmond,Adelaide   
23               0                  Ryerson,Garden District   
24               0                           St. James Town   
25               0               St. James Town,Cabbagetown   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
8                   Café                 Hotel           Coffee Shop   
15                  Café           Coffee Shop                 Hotel   
23           Coffee Shop                  Café        Cosmetics Shop   
24                  Café           Coffee Shop            Restaurant   
25                  Park                  Café   Japanese Restaurant   

        4th Most Common Venue  
8                  Steakhouse  
15                    Theater  
23  Middle Eastern Restaurant  
24                      Hotel  
25                      Diner

Cluster 1

In [55]:
target_venues_sorted.loc[target_venues_sorted['Cluster Labels'] == 1]

Cluster Labels                                Neighborhood  \
3                1     Chinatown,Kensington Market,Grange Park   
4                1                                    Christie   
10               1  Exhibition Place,Brockton,Parkdale Village   
32               1                The Junction South,High Park   
33               1                     Trinity,Little Portugal   
35               1               University of Toronto,Harbord   
37               1           Yorkville,The Annex,North Midtown   

   1st Most Common Venue          2nd Most Common Venue  \
3                   Café  Vegetarian / Vegan Restaurant   
4      Korean Restaurant                           Café   
10                  Café                    Coffee Shop   
32                   Bar                           Café   
33                   Bar                           Café   
35                  Café                         Bakery   
37                  Café  Vegetarian / Vegan Restaurant   

            3rd Most Common Venue   4th Most Common Venue  
3                             Bar      Mexican Restaurant  
4                   Grocery Store             Coffee Shop  
10                            Bar  Furniture / Home Store  
32                    Coffee Shop       Convenience Store  
33                         Bakery             Coffee Shop  
35  Vegetarian / Vegan Restaurant                     Bar  
37             Italian Restaurant             Coffee Shop

Cluster 2

In [56]:
target_venues_sorted.loc[target_venues_sorted['Cluster Labels'] == 2]

Cluster Labels                   Neighborhood 1st Most Common Venue  \
13               2  India Bazaar,The Beaches West     Indian Restaurant   
20               2          Roncesvalles,Parkdale           Coffee Shop   
31               2                    The Beaches                   Pub   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  
13           Coffee Shop                  Café                 Beach  
20           Pizza Place                   Bar                  Café  
31           Pizza Place           Coffee Shop                 Beach

### Result

Firstly, by using k-means clustering method the neighborhoods are clustered according to their frequency of having coffee shop. Than the cluster having the minimum frequency is selected for further analysis. In this cluster, neighborhoods are sorted with respect to the 4 common venues in those neighborhoods. Again k-means clustering method is applied. According to the result, neighborhoods in cluster 0 have their most common values Cafes, and Coffee Shops. Also there are most common venues which can increase the customers of coffee shop such as hotel, park, theatre. So that the neighborhoods in cluster 0 are suggested locations to open a coffee shop.